README.md
# fluent-logger for Dartlang

fluent-logger implementation for Dart inspired by [fluent-logger-perl](https://github.com/fluent/fluent-logger-perl)

UNDER DEVELOPMENT.

If found bug or issues, please contribute to [the fluent-logger-dart repository!](https://github.com/ichigotake/fluent-logger-dart/issues)

## Usage

They currently only work on platforms where `dart:io` is available.

``` dart
FluentLogger logger = new FluentLogger(host: '127.0.0.1', port: 24224, timeout: 300);
logger.post("test.greet", {"greeting": "Good afternoon :)"})
  .then((e) => 'success!')
  .catchError((e) => 'failed...')
  .whenComplete((action) => logger.destroy());
```

## TODO

- Retries posting message when failed to post to `fluentd`.

## License

MIT License

## See also

- [https://github.com/ichigotake/fluent-logger-dart](https://github.com/ichigotake/fluent-logger-dart)
- [https://pub.dartlang.org/packages/fluent_logger](https://pub.dartlang.org/fluent_logger)
- [Fluentd | Open Source Log Management](http://www.fluentd.org/)
- [https://github.com/fluent/fluentd](https://github.com/fluent/fluentd)


CHANGELOG.md
# Changelog

## v0.1.4, v0.1.3
- Misstake on publishing

## v0.1.2
- Change owner email address

## v0.1.1
- Write document for working on platform with `dart:io`

## v0.1.0

- Implement FluentLogger\#post
- Implement retry connection when connection failed


test/post-test.dart
library LiveTest;

import 'dart:convert' show JSON;
import 'dart:async';
import 'dart:io' show Socket, Process, ProcessResult;
import 'package:unittest/unittest.dart';
import '../lib/fluent_logger.dart';

/**
 *
 * Before run `fluentd` on localhost.
 *
 *    <source>
 *      type forward
 *      port 24224
 *    </source>
 *    <match test.*>
 *      type file
 *      path $path_to_log_file
 *    </match>
 *
 * And run this file.
 */
main() {
  FluentLogger logger = new FluentLogger(host: '127.0.0.1', port: 24224, timeout: new Duration(milliseconds: 300));

  test("Post an message", (){
    return logger.post("test.greet", {"greeting": "Hello world!"});
  });

  test("Post an message", (){
    return logger.post("test.greet", {"greeting": "Good afternoon :)"});
  });

  test("Message must be type as Map", (){
    logger.post("test.greet", "Hello world!")
      .then((e) => fail('failed'))
      .catchError((e) => expectAsync(e));
  });

  test("Destroy socket", (){
    return logger.destroy();
  });

  test("Post an message after destroy socket", (){
    return logger.post("test.greet", {"greeting": "Good night..."});
  });

  test("Too many post few times", (){
    Completer completer = new Completer();
    for (int i=0; i<1000; i++) {
      new Future((){
        logger.post("test.greet", {"greeting": "Storm greeting! " + i.toString()});
      });
    }
    completer.complete();
    return completer.future;
  });

  test("Destroy socket", (){
    return logger.destroy();
  });


}


lib/fluent-logger/fluent-logger.dart
library FluentLogger;

import 'dart:io' show Socket, sleep;
import 'dart:async' show Completer, Future;
import 'dart:convert' show JsonEncoder;

class FluentLogger {

  Socket _socket;
  var _host;
  int _port;
  Duration _timeout;
  int _maxRetryCount;
  Duration _retryInterval;
  String _tagPrefix;
  JsonEncoder _packer;

  // For socket buffer
  String buffer = "";
  bool sending = false;

  FluentLogger({host, int port, Duration timeout, String tagPrefix, Duration retryInterval, int maxRetryCount}) {
    this._host = host != null ? host : '127.0.0.1';
    this._port = port != null ? port : 24224;
    this._timeout = timeout != null ? timeout : new Duration(milliseconds: 500);
    this._maxRetryCount = maxRetryCount != null ? _maxRetryCount : 12;
    this._retryInterval = retryInterval != null ? retryInterval : new Duration(milliseconds: 500);
    this._tagPrefix = tagPrefix;
    this._packer = new JsonEncoder();
  }

  Future<Socket> connect() {
    return Socket.connect(_host, _port)
      .timeout(_timeout)
      .then((Socket socket){
        _socket = socket;
      })
      .catchError((e){
        if (_socket != null) {
          _socket.destroy();
        }
        print("Cannot connect socket: ${e}");
      });
  }

  Future post(String tag, Map message, {int timestamp}) {
    Completer completer = new Completer();
    if (!(message is Map)) {
      completer.completeError("Message must be type as Map.");
      return completer.future;
    }
    String sendTag = _tagPrefix != null ? "${_tagPrefix}.${tag}" : tag;
    int sendTimestamp = timestamp != null ? timestamp : new DateTime.now().millisecondsSinceEpoch/1000;
    var sendData = _packer.convert([sendTag, sendTimestamp, message]);
    buffer += sendData + "\n";
    if (sending) {
      completer.complete();
      return completer.future;
    }
    sending = true;

    if (_socket == null) {
      connect()
        .then((Socket socket) => _send(completer, buffer))
        .catchError((e){
          print("Connect error: ${e}");
          sending = false;
          completer.complete(_socket);
        });
    } else {
      _send(completer, buffer);
    }
    return completer.future;
  }

  _send(Completer completer, sendData) {
    sending = true;
    int retries = 1;
    while (true) {
      sleep(_retryInterval);
      if (retries++ > _maxRetryCount) {
        break;
      }
      _socket.write(sendData);
      sending = false;
      buffer = "";
      break;
    }
    completer.complete(_socket);
  }

  void destroy() {
    if (_socket != null) {
      _socket.destroy();
      _socket = null;
    }
  }

}

// Which Error or Exception?
class ArgumentTypeException implements Exception {
  String message;

  ArgumentTypeException(this.message);

  @override
  String toString() => "ArgumentError: $message";

}


lib/fluent_logger.dart
/**
 * Fluent logger for Dart.
 *
 * For information on installing and importing this library, see the
 * [fluent_logger package on pub.dartlang.org]
 * (http://pub.dartlang.org/packages/fluent_logger).
 */
library fluent_logger;

import 'dart:io' show Socket, sleep;
import 'dart:async' show Completer, Future;
import 'dart:convert' show JsonEncoder;

/**
 * Use a [FluentLogger] to logging messages to fluentd.
 */
class FluentLogger {

  Socket _socket;
  var _host;
  int _port;
  Duration _timeout;
  int _maxRetryCount;
  Duration _retryInterval;
  String _tagPrefix;
  JsonEncoder _packer;

  // For socket buffer
  String _buffer = "";
  bool sending = false;

  FluentLogger({host, int port, Duration timeout, String tagPrefix, Duration retryInterval, int maxRetryCount}) {
    this._host = host != null ? host : '127.0.0.1';
    this._port = port != null ? port : 24224;
    this._timeout = timeout != null ? timeout : new Duration(milliseconds: 500);
    this._maxRetryCount = maxRetryCount != null ? _maxRetryCount : 12;
    this._retryInterval = retryInterval != null ? retryInterval : new Duration(milliseconds: 500);
    this._tagPrefix = tagPrefix;
    this._packer = new JsonEncoder();
  }

  /**
   * Connect to fluentd socket.
   */
  Future<Socket> connect() {
    return Socket.connect(_host, _port)
      .timeout(_timeout)
      .then((Socket socket){
        _socket = socket;
      })
      .catchError((e){
        if (_socket != null) {
          destroy();
        }
        print("Cannot connect socket: ${e}");
      });
  }

  /**
   * Post log message to fluentd.
   */
  Future post(String tag, Map message, {int timestamp}) {
    Completer completer = new Completer();
    if (!(message is Map)) {
      completer.completeError("Message must be type as Map.");
      return completer.future;
    }
    String sendTag = _tagPrefix != null ? "${_tagPrefix}.${tag}" : tag;
    int sendTimestamp = timestamp != null ? timestamp : new DateTime.now().millisecondsSinceEpoch/1000;
    var sendData = _packer.convert([sendTag, sendTimestamp, message]);
    _buffer += sendData + "\n";
    if (sending) {
      completer.complete();
      return completer.future;
    }
    sending = true;

    if (_socket == null) {
      connect()
        .then((Socket socket) => _send(completer, _buffer))
        .catchError((e){
          sending = false;
          completer.complete(_socket);
        });
    } else {
      _send(completer, _buffer);
    }
    return completer.future;
  }

  /**
   * Send logging message to fluentd.
   */
  _send(Completer completer, sendData) {
    sending = true;
    int retries = 1;
    while (true) {
      sleep(_retryInterval);
      if (retries++ > _maxRetryCount) {
        break;
      }
      _socket.write(sendData);
      sending = false;
      _buffer = "";
      break;
    }
    completer.complete(_socket);
  }

  /**
   * Destroy socket connection.
   */
  void destroy() {
    if (_socket != null) {
      _socket.destroy();
      _socket = null;
    }
  }

}

// Which Error or Exception?
class ArgumentTypeException implements Exception {
  String message;

  ArgumentTypeException(this.message);

  @override
  String toString() => "ArgumentError: $message";

}
